### ISSUE TITLE
Haskell mode Evil indentation issue

### ISSUE THREAD

**Issue by Eilie** [syl20bnr/spacemacs#5010]
#### System Info
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.105.9
- Spacemacs branch: master (rev. 0562f05)
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(haskell)
```
#### Steps to reproduce
1. Open .hs file.
2. Type `foo :: a`.
3. Press `ESC`.
4. Press `o`.

#### Expected
Cursor is moved under "f" letter.

#### Actual
Cursor is moved under "a" letter.

**Comment by StreakyCobra** [comment]
I think it's a duplicate of #621, isn't it?

**Comment by d12frosted** [comment]
It's not a duplicate of #621 for sure, but both issues are common - haskell-mode and vim editing style don't play well together. Most of things work as expected, but there are several exceptions. 

I am not using vim editing style, so I don't know what exactly `o` is, but from what I see (via description system and results I have) it's almost the same as pressing `C-e` (go to end of line), `return` (new line) and `tab` to indent. Your problem actually is in the last `tab`.

```
foo :: a
|
```

If you press `tab` in that situation it will move your cursor to the position of `a`. And this is expected behaviour of `haskell-indentation`. 

So in my opinion you have three possible solutions.

1. Find a command similar to `o`, but that doesn't indent (I believe that it should exist). 
2. Implement that command yourself (or ask someone) and bind it to `o` in haskell-mode for normal state (or how is it called?). 
3. Open an issue in `haskell-mode` [issue tracker](https://github.com/haskell/haskell-mode/issues). 

If you need help with second proposed solution - just let us know. Or you could join [Spacemacs Gitter Chat](https://gitter.im/syl20bnr/spacemacs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) - there are plenty of users that will be glad to help you :)

**Comment by AlejandroCatalina** [comment]
Actually, wouldn't `newline` alone be the command you are looking for?

**Comment by d12frosted** [comment]
@AlejandroCatalina ah... didn't know that simple return in normal mode doesn't break line. :smile: Makes sense, yeah. Well, as I said, I don't use vim editing style, so don't know how it works :D 

So yeah, then it falls into one simple solution - use return :D

**Comment by AlejandroCatalina** [comment]
Yeah using return should not indent. We might define a new command to go to end of the line before entering return but i think that could basically work, unless more specifications provided.

**Comment by d12frosted** [comment]
Ah. And the thing I forgot about `o` - it also switches from normal mode to insert mode. So probably OP would like to have a command that makes `return` from normal mode and switches to insert mode.

**Comment by AlejandroCatalina** [comment]
Yeah, that'd be nice too, @Eilie would this solve the issue?

**Comment by StreakyCobra** [comment]
`o` is `evil-open-bellow`. There is already an `haskell-mode` [issue](https://github.com/haskell/haskell-mode/issues/896) for `evil-open-above`, so as they are related this one should probably be reported on the same issue.

**Comment by d12frosted** [comment]
Ok. Not I totally feel myself silly :D 
Good to know that there is a similar issue for that :D

**Comment by jb55** [comment]
This issue is brutal, compounded by that fact you have to press TAB twice for it to fix itself. So `o` now becomes `o<TAB><TAB>` ugh.

**Comment by d12frosted** [comment]
you could also press `S-TAB` :)

**Comment by jb55** [comment]
I get <backtab> is undefined

**Comment by jb55** [comment]
I guess the evil version of that is ctrl-w, which is slightly more bearable

**Comment by jb55** [comment]
actually not quite, that makes it go to the start of the line, which doesn't work for let indents and such :[

**Comment by fice-t** [comment]
Not exactly an Evil/spacemacs guy, but isn't this just because `evil-auto-indent` is `t` in haskell buffers?

```elisp
(when evil-auto-indent
    (indent-according-to-mode))
```
It's right at the end of the source for `evil-open-above` and `evil-open-below`.

I'm not sure why this is considered an issue only in Haskell.

**Comment by StreakyCobra** [comment]
> I'm not sure why this is considered an upstream issue.

Because the similar issue about `evil-open-above` (https://github.com/syl20bnr/spacemacs/issues/3162) was reported upstream by on of the upstream developers (https://github.com/haskell/haskell-mode/issues/896) himself, so I suppose it's an upstream issue.

> but isn't this just because `evil-auto-indent` is `t` in haskell buffers?

It is set to `t` in `python-mode` too and I don't notice this behaviour.

**Comment by fice-t** [comment]
Now I'm not sure what the expected behaviour is supposed to be, since I appear to see the same behaviour in Python. Namely:

1. Type `def foo(object):`
2. Press `ESC`
3. Press `o`

This does indent like it does in Haskell. An example where the indent in `python-mode` isn't like what you might expect/want:

```python
while True:
    do_something()
    do_more_things()

do_other_stuff()
^
```
If you press `S-o` with the cursor where the `^` is pointing, then you'll get:

```python
while True:
    do_something()
    do_more_things()

    _
do_other_stuff()
```
With the cursor now being the indented `_`.

So to me, it still seems like in both Haskell and Python, `evil-auto-indent` should be nil. From the issues I've seen here, it doesn't seem like people want auto indentation after using `evil-open-x` in Haskell. Setting it to nil would give these people the expected behaviours that they described.

All that said, there are issues with Haskell indentation all over the place, but I don't see how it deeply relates specifically to `evil-open-x`.

**Comment by StreakyCobra** [comment]
This variable just define if auto-indent should be done after entering the insert state. It's the mode which is responsible to indent smartly. For instance in Python:

```python
while True:
    do_something()
    do_more_things()

do_other_stuff()
^
```

will indent it because there is no sign that the previous block is finished, but

```python
while True:
    do_something()
    do_more_things()
    break

do_other_stuff()
^
```

will not indent it as the previous statement have a returning statement (break, continue, return, pass, …). Also if you do this on the first function/block of the file it will not be indented.

On the other hand if we set this variable to `nil`, this wrong behavior will happen:

```python
while True:
    do_something()
    ^
    do_more_things()

do_other_stuff()
```

pressing `o` from this position should indent it as we are still in the block, but instead the output is:


```python
while True:
    do_something()

^
    do_more_things()

do_other_stuff()
```

which is probably probably even worse. So the current behavior is maybe not perfect, but it is still the better one. Also it is still possible to press `TAB` to switch between the different indentations of the line that just get opened.

Anyway, we are not setting `evil-auto-indent` anywhere in Spacemacs, and we probably shouldn't touch it. It's an user preference, and if someone don't like it, it can easily be switched per mode with:

```elisp
(add-hook 'python-mode-hook (lambda () (setq evil-auto-indent nil)))
(add-hook 'haskell-mode-hook (lambda () (setq evil-auto-indent nil)))
```

in `user-config`, or generally with:

```elisp
(setq-default evil-auto-indent nil)
```

in `user-init`.

----

To go back to Haskell, the function which is called every time to set the indentation is `indent-line-function` value, which is set to `haskell-indentation-indent-line`. Calling this function on the example given above:

```haskell
foo :: a
|
```

will indent it wrongly to:

```haskell
foo :: a
       |
```

The `haskell-indentation-indent-line` function being defined in `haskell-mode`, it means that's an upstream issue from my point of view.

**Comment by fice-t** [comment]
To me it seems that it shouldn't auto-indent if there's any ambiguity (as in the case with no `break`), but perhaps that's just me. I do agree that `haskell-mode` could use a lot better heuristics to figure out proper indentation, though as far as I'm aware, due to the structure of Haskell this could be difficult for some cases.

For example, that 'wrong' indent could be considered a valid secondary indent as
```haskell
foo :: Int
       -> Int
foo = succ
```
is valid Haskell. It just so happens that in most cases you do not want to indent at that line.

I looked at the code a bit and there might be something that may work. Again, I don't really use Evil, so others will have to test this out. But replace your current definition of `haskell-indentation-indent-line` with:

```haskell
(defun haskell-indentation-indent-line ()
  "Indent current line, cycle though indentation positions.
Do nothing inside multiline comments and multiline strings.
Start enumerating the indentation points to the right.  The user
can continue by repeatedly pressing TAB.  When there is no more
indentation points to the right, we switch going to the left."
  (interactive)
  ;; try to repeat
  (when (not (haskell-indentation-indent-line-repeat))
    (setq haskell-indentation-dyn-last-direction nil)
    ;; parse error is intentionally not cought here, it may come from
    ;; `haskell-indentation-find-indentations', but escapes the scope
    ;; and aborts the opertaion before any moving happens
    (let* ((cc (current-column))
           (ci (haskell-indentation-current-indentation))
           (inds (save-excursion
                   (move-to-column ci)
                   (or (haskell-indentation-find-indentations)
                       '(0))))
           (valid (memq ci inds))
           (cursor-in-whitespace (< cc ci)))

      (if (and valid cursor-in-whitespace)
          (move-to-column ci)
        (haskell-indentation-reindent-to
         (haskell-indentation-next-indentation (if (bound-and-true-p evil-mode) (1- ci) ci) inds 'nofail)
         cursor-in-whitespace))
      (setq haskell-indentation-dyn-last-direction 'right
            haskell-indentation-dyn-last-indentations inds))))
```
The only difference is the `bound-and-true-p` that will include the indentation level that you're currently on only if you're using evil.

This makes it like `python-mode` where you may have to TAB twice to get it moving, which is why for non-evil users it probably isn't a nice addition.

**Comment by StreakyCobra** [comment]
Ok. As I wrote upstream, the discussion belongs there between interested people

**Comment by jb55** [comment]
Should this be closed then since it's a nofix?

**Comment by StreakyCobra** [comment]
It can stay open for now. This will allow people to see there is an issue and that it has been reported upstream :smiley:

**Comment by d12frosted** [comment]
> This will allow people to see there is an issue

So they don't open similar issues again :) 

P. S. nice conversation :+1:

**Comment by Eilie** [comment]
@fice-t Good job, thank you! But how can I use it within spacemacs, so far I've managed to make this fix work by evaluating it in *scratch* but  placing it in my `.spacemacs` doesn't work.

**Comment by fice-t** [comment]
That'd probably be because the file that contains the function `haskell-indentation` is loaded after your .spacemacs. There's a few options: require the file before defining that function, wrap it in an `with-eval-after-load`, or manually edit that file with the new function (it would be overwritten on a new version of `haskell-mode`). After a fix has been merged you can remove anything you put in your init.

Look [here](https://github.com/haskell/haskell-mode/issues/896#issuecomment-185945576) for a better version of the function, though.

**Comment by kiela** [comment]
@Eilie can you contact me over an email?

**Comment by d12frosted** [comment]
Since it's a duplicate, closing in favour of https://github.com/syl20bnr/spacemacs/issues/5010. If someone thinks that it should be left open - let me know and I'll reopen. 

P. S. thanks everyone for looking into this issue.

**Comment by bennofs** [comment]
@d12frosted closing in favor of #5010? But that is this very issue!

**Comment by afcady** [comment]
I think they meant it's closed in favor of #621
